File: fan_lsqt/test_hamiltonian_01.jl
using Printf
using SparseArrays

using Infiltrator

function gen_idx_xy2ip(Nx, Ny)
    idx_xy2ip = zeros(Int64, Nx, Ny)
    ip = 0
    for j in 1:Ny, i in 1:Nx
        ip += 1
        idx_xy2ip[i,j] = ip
    end
    return idx_xy2ip
end


function test_main()

    Nx = 10
    Ny = 10
    W = 1.0


    idx_xy2ip = gen_idx_xy2ip(Nx, Ny)

    Npoints = Nx * Ny # total number of sites

    row_H = zeros(Int64, Npoints*4 - Nx*2) # the row indices for H
    col_H = zeros(Int64, Npoints*4 - Nx*2) # the column indices for H
    Hij = -ones(ComplexF64, Npoints*4 - Nx*2) # nonzero Hamiltonian elements

    row_V = zeros(Int64, Npoints*2) # row indices for V
    col_V = zeros(Int64, Npoints*2) # column indices for V
    Vij = zeros(ComplexF64, Npoints*2) # nonzero velocity matrix elements

    row_U = 1:Npoints # row and column indices for U
    col_U = 1:Npoints
    Uij = 1.234*ones(Npoints) #np.random.uniform(-W * 0.5, W * 0.5, N) # on-site potentials

    count_H = 0
    count_V = 0

    for j in 1:Ny, i in 1:Nx
        #
        # (0) # get the index of the center site
        idx_center = idx_xy2ip[i,j]
        #
        # (1) consider the left neighbor (periodic boundary)
        ii = i - 1
        if ii <= 0
            ii += Nx # wrap
        end
        idx_left = idx_xy2ip[ii,j]
        #
        count_H += 1    
        row_H[count_H] = idx_center
        col_H[count_H] = idx_left
        #
        count_V += 1
        row_V[count_V] = idx_center
        col_V[count_V] = idx_left
        Vij[count_V] = im
        #
        # (2) consider the right neighbor (periodic boundary)
        ii = i + 1
        if ii >= Nx + 1
            ii -= Nx # wrap
        end
        idx_right = idx_xy2ip[ii,j]
        #
        count_H += 1
        row_H[count_H] = idx_center
        col_H[count_H] = idx_right
        #
        count_V += 1
        row_V[count_V] = idx_center
        col_V[count_V] = idx_right
        Vij[count_V] = -im
        #
        # (3) consider the upper neighbor (open boundary)
        @printf("\n -- i=%3d j=%3d idx_center=%d\n", i, j, idx_center)
        @printf("Periodic BC: idx_left=%3d idx_right=%3d\n" , idx_left, idx_right)
        if j < Ny
            idx_up = idx_xy2ip[i,j+1]
            count_H += 1
            row_H[count_H] = idx_center
            col_H[count_H] = idx_up
            @printf("Open BC: idx_up=%3d\n", idx_up)
        end
        #
        # (4) consider the down neighbor (open boundary)
        if j > 1
            idx_down = idx_xy2ip[i,j-1]
            count_H += 1
            row_H[count_H] = idx_center
            col_H[count_H] = idx_down
            @printf("Open BC: index_down=%3d\n", idx_down)
        end
    end

    H = sparse( row_H, col_H, Hij )
    U = sparse( row_U, col_U, Uij )
    H = H + U
    V = sparse( row_V, col_V, Vij )

    @infiltrate

end # function

test_main()

File: fan_lsqt/test_hamiltonian_01.py
import numpy as np
from mod_lsqt_01 import *

Nx = 50000
Ny = 2
W = 1.0

N = Nx * Ny; # total number of sites

row_H = np.zeros(N * 4 - Nx * 2) # the row indices for H
col_H = np.zeros(N * 4 - Nx * 2) # the column indices for H
Hij = -np.ones(N * 4 - Nx * 2, complex) # nonzero Hamiltonian elements

row_V = np.zeros(N * 2) # row indices for V
col_V = np.zeros(N * 2) # column indices for V
Vij = np.zeros(N * 2, complex) # nonzero velocity matrix elements

row_U = np.arange(0, N) # row and column indices for U
Uij = 1.234*np.ones(N) # np.random.uniform(-W * 0.5, W * 0.5, N) # on-site potentials

count_H = 0 # number of nonzero H elements
count_V = 0 # number of nonzero V elements

for nx in range(Nx):
    for ny in range(Ny):
        # (0) # get the index of the center site
        index_center = find_index(nx, ny, Ny)
        # (1) consider the left neighbor (periodic boundary)
        index_left = find_index((nx - 1) % Nx, ny, Ny)
        row_H[count_H] = index_center
        col_H[count_H] = index_left
        count_H += 1
        row_V[count_V] = index_center
        col_V[count_V] = index_left
        Vij[count_V] = 1j
        count_V += 1
        # (2) consider the right neighbor (periodic boundary)
        index_right = find_index((nx + 1) % Nx, ny, Ny)
        row_H[count_H] = index_center
        col_H[count_H] = index_right
        count_H += 1
        row_V[count_V] = index_center
        col_V[count_V] = index_right
        Vij[count_V] = -1j
        count_V += 1
        # (3) consider the upper neighbor (open boundary)
        #print("\n -- nx=%3d ny=%3d index_center=%d" % (nx, ny, index_center))
        #print("Periodic BC: index_left=%3d index_right=%3d" % (index_left, index_right))
        if ny < Ny - 1:
            index_up = find_index(nx, (ny + 1), Ny)
            row_H[count_H] = index_center
            col_H[count_H] = index_up
            count_H += 1
            #print("Open BC: index_up=%3d" % (index_up))
        #else:
        #    print("No index up")
        # (4) consider the down neighbor (open boundary)
        if ny > 0:
            index_down = find_index(nx, (ny - 1), Ny)
            row_H[count_H] = index_center
            col_H[count_H] = index_down
            count_H += 1
            #print("Open BC: index_down=%3d" % (index_down))
        #else:
        #    print("No index down")


H = sparse.csr_matrix((Hij, (row_H, col_H)), shape = (N, N))
U = sparse.csr_matrix((Uij, (row_U, row_U)), shape = (N, N))
H = H + U
V = sparse.csr_matrix((Vij, (row_V, col_V)), shape = (N, N))

#Udense = U.todense()
#Vdense = V.todense()
#Hdense = H.todense()
